### PR TITLE
Added pushState fallback for IE 9 browser. fixes #6506

### LINKED ISSUE #6506: Angular 2, IE9 & pushState issue!
Hi,

I'm having some issues with Angular 2, pushState & IE9 as I've documented initially in https://groups.google.com/forum/#!topic/angular/TdCt6KyhwI8 before carrying out much more investigation!

I've read https://github.com/angular/angular/issues/5792 and attempted to apply the history.js script with mixed results!

I have had to modify angular2.dev.js locally to return window.History instead of window.history (not a long term solution really!)

In Everything but IE9 things work as expected using path location mode - in IE9 history.js takes over and injects urls to look like http://my.app.com/#./dashboard

But those hashtag paths are then not recognised by the angular router..!!

So I turn on the hash location strategy and that works fine in chrome etc but in IE history.js kicks in and alerts the following..

>> History.js does not support states with fragment-identifiers (hashes/anchors). 

The root issue is found in the following code in the router..

```
    PlatformLocation.prototype.pushState = function(state, title, url) {
        console.log('pushState line 2022');
      this._history.pushState(state, title, url);
      console.log('after push line 2026');
    };
```
If I comment out the pushState line then everything works.. but of course the url is not updated in the browser window :(

It feels a bit like a catch-22 situation!!

a) Is this a bug??
b) Does the Angular team have a compatible IE9 routing solution?

Help!!!

### PR DESCRIPTION
**Pull request by sasidhar**
- **The commit message follows our guidelines:**
  https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit-message-format
- **Tests for the changes have been added (for bug fixes / features)**
  No tests added, as the changes are only for IE9 browsers where `pushState` is not available
- **Docs have been added / updated (for bug fixes / features)**
  Not updated. As we are already mentioning that Angular 2 supports IE9, and as this is a bug fix, I don't see any need for updating Docs
- **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
  Only code changes
- **What is the current behavior?** (You can also link to an open issue here)
  #6506
- **What is the new behavior (if this is a feature change)?**
  This should work in IE9 without any `pushState` issues
- **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
  No

### CODE REVIEW

File: modules/angular2/src/router/location/browser_platform_location.ts [hunk 1]
```diff
@@ -45,11 +46,19 @@ export class BrowserPlatformLocation extends PlatformLocation {
   set pathname(newPath: string) { this._location.pathname = newPath; }
 
   pushState(state: any, title: string, url: string): void {
-    this._history.pushState(state, title, url);
+    if(isPresent(this._history.pushState)) {
```
Comment (mhevery) [inline modules/angular2/src/router/location/browser_platform_location.ts]: I am sorry but this code will not work in Dart. In dart you need to do something like this: https://api.dartlang.org/stable/1.15.0/dart-html/History/supportsState.html

This means that the check needs to be in a facade implemented twice, once in TS and once in Dart. (see `isPresent` as an example of how to do that.)

**Comment by mhevery** [comment]
Thanks for your contribution. Would you be willing to do the work which is needed to make this pass on Dart?

**Comment by sasidhar** [comment]
@mhevery Definitely I will try. 

As per your comments
- We add `supportsState` method to `lang.ts` which uses `isPresent` internally to test whether state is supported.
- We add `supportsState` method to `lang.dart` which uses core dart static method `supportsState`
- We use `supportsState` method in `browser_platform_location.ts` instead of `isPresent`

Am I correct?

**Comment by mhevery** [comment]
Yes!

**Comment by sasidhar** [comment]
@mhevery To add `supportsState` to `lang.ts`, it seems, we need to import `DOM` from `angular2/src/platform/dom/dom_adapter` to access History object. But is it suggestible to import the same in `lang.ts`

I am planning to add `supportsState` method to `lang.ts` as following
```
export function supportsState(): boolean {
  return isPresent(DOM.getHistory.pushState);
}
```

**Comment by mhevery** [comment]
@sasidhar Instead putting it in `lang.ts` create a local `location.ts` and `location.dart` next to the `browser_platform_location` and put the two implementations there.

**Comment by sasidhar** [comment]
@mhevery We already have a `location.ts` next to `browser_platform_location.ts`. So, I used `history.ts` and `history.dart`. But seems this commit has some conflicts.

Please let me know what needs to be done from my side to fix these conflicts.

**Comment by ggaulard** [comment]
Just to let you know, This stuff is really needed 👍 
Thanks

**Comment by poke** [comment]
Is there anything that prevents this fix from being merged?

**Comment by martinth** [comment]
@mhevery is there anything that prevents this from being merged? 

I'm asking because I just ran into this issue when using Angular 2 inside an Microsoft Office Addin (which does some weird shit and patches out `pushState`) so this code would also help there.

**Comment by poke** [comment]
Thank you for the fix!

**Pull request closed without merging**